### PR TITLE
floating action buttons

### LINKED ISSUE #1714: [F100] Floating Action Buttons
## Rationale

A floating action button menu is a common design paradigm. There is currently no control in Xamarin.Forms that will provide this functionality on all platforms.

## Implementation

Add new control `FloatingActionButton`:

```csharp
public static readonly BindableProperty ColorProperty = BindableProperty.Create("Color", typeof(Color), typeof(FloatingActionButton), Color.Default);

public static readonly BindableProperty ImageSourceProperty = BindableProperty.Create("ImageSource", typeof(ImageSource), typeof(FloatingActionButton), null);

public static readonly BindableProperty CommandProperty = BindableProperty.Create("Command", typeof(ICommand), typeof(FloatingActionButton), null);

public static readonly BindableProperty CommandParamenterProperty = BindableProperty.Create("CommandParameter", typeof(object), typeof(FloatingActionButton), null);

public static readonly BindableProperty SizeProperty = BindableProperty.Create("Size", typeof(FloatingActionButtonSize), typeof(FloatingActionButton), FloatingActionButtonSize.Normal);

public event EventHandler Clicked;

public enum FloatingActionButtonSize
{
    Normal,
    Mini
}
```

Renderers will need to be created accordingly. 

### Android

- Use the platform's built-in `FloatingActionButton` control.

### iOS

- Create custom button.

### UWP

- Create custom button.

## Implications for CSS

`FloatingActionButton` will support `color`.

## Backward Compatibility

Since we are adding a new control, there should be no compatibility problems at all.

## Difficulty : High
This is a brand new control with commands and events, and there is no corresponding platform control for UWP or iOS. Consistency and stability will be very important.

### PR DESCRIPTION
**Pull request by sichy**
### Description of Change ###

IOS and Droid implementation for FloatingActionButton a rounded button with icon and Command. On Android implemented as native FloatingActionButton.

### Issues Resolved ###

- fixes #1714

### API Changes ###

Added:
 - control FloatingActionButton

### Platforms Affected ###

- Core/XAML (all platforms)
- iOS
- Android

### Testing procedure ###

1. Dynamic View Gallery has the FloatingActionButton available, so you can play with all the properties.
2. FloatingActionButton Gallery has similar tests.
3. Sandbox has a single FAB.
4. XamStore >> List (tab) has a fun popup FAB menu with animation.

fab

### PR Checklist ###

- [x] Has automated tests, added the test page to control forms gallery
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by sichy** [comment]
@rcunniffe unfortunately cannot tell I do not have access to the "details" section of the failed parts :/

**Comment by rcunniffe** [comment]
@samhouts what's the protocol here then?

**Comment by samhouts** [comment]
```Xamarin.Forms.Platform.cs(136,21): Error CS0246: The type or namespace name 'FloatingActionButtonRenderer' could not be found (are you missing a using directive or an assembly reference?)```

File: Xamarin.Forms.Platform.Android/packages.config [hunk 1]
```diff
@@ -1,20 +1,20 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="Xamarin.Android.Arch.Core.Common" version="1.0.0" targetFramework="monoandroid81" />
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/packages.config]: Was this necessary?

**Comment by samhouts** [comment]
@sichy I realize it has been a month since this was submitted. Let us know if you have time to continue work on it, or if you would like for us to take it over from here. Thank you!!!!

File: Xamarin.Forms.Core/FloatingActionButton.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+using System.Windows.Input;
+using Xamarin.Forms.Internals;
+using Xamarin.Forms.Platform;
+
+namespace Xamarin.Forms
+{
+	[RenderWith(typeof(_FloatingActionButtonRenderer))]
+	public class FloatingActionButton : View, IButtonController
+	{
+		public static readonly BindableProperty ImageSourceProperty = BindableProperty.Create("ImageSource", typeof(ImageSource), typeof(FloatingActionButton), null);
```
Comment (jassmith) [inline Xamarin.Forms.Core/FloatingActionButton.cs]: use nameof()

File: Xamarin.Forms.Core/FloatingActionButton.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+			if (IsEnabled == true)
+			{
+				Pressed?.Invoke(this, EventArgs.Empty);
+			}
+		}
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
+		public void SendReleased()
+		{
+			if (IsEnabled == true)
```
Comment (jassmith) [inline Xamarin.Forms.Core/FloatingActionButton.cs]: if (IsEnabled)

File: Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+using AMotionEventActions = Android.Views.MotionEventActions;
+using Object = Java.Lang.Object;
+using Xamarin.Forms.Internals;
+using System.Threading.Tasks;
+using Android.Graphics.Drawables;
+using Android.Content.Res;
+
+namespace Xamarin.Forms.Platform.Android
+{
+	public class FloatingActionButtonRenderer : ViewRenderer<FloatingActionButton, AFloatingButton>, AView.IOnAttachStateChangeListener
```
[the remark this answers is not in the document]
Comment (jassmith) [reply Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs]: Agreed, fast renderer only

**Comment by ianvink** [comment]
Just a kind thank-you for all those working on this feature. I wish I have the skillz to contribute code so I simply contribute gratitude.

**Comment by samhouts** [comment]
@sichy Just a friendly ping! We're going to pick this up and finish it up next week, unless you've already planned to make changes to it. We'll assume that no response is approval for us to continue the work. Thank you!

File: Xamarin.Forms.ControlGallery.iOS/AppDelegate.cs [hunk 5]
```diff
@@ -164,7 +164,7 @@ public override bool FinishedLaunching(UIApplication uiApplication, NSDictionary
 				}
 			};
 
-			if (App.IOSVersion == 11)
+			if (App.IOSVersion >= 11)
```
Comment (samhouts) [inline Xamarin.Forms.ControlGallery.iOS/AppDelegate.cs]: Not sure where this came from

File: Stubs/Xamarin.Forms.Platform.cs [hunk 6]
```diff
@@ -154,6 +154,11 @@ internal class _PageRenderer { }
 	[RenderWith (typeof (PhoneMasterDetailRenderer))]
 #endif
 	internal class _MasterDetailPageRenderer { }
+
+#if  !TIZEN4_0
```
Comment (samhouts) [inline Stubs/Xamarin.Forms.Platform.cs]: WPF? GTK?

File: Xamarin.Forms.Platform.Android/Xamarin.Forms.Platform.Android.csproj [hunk 7]
```diff
@@ -18,7 +18,7 @@
     <AndroidResgenFile>Resources\Resource.Designer.cs</AndroidResgenFile>
     <GenerateSerializationAssemblies>Off</GenerateSerializationAssemblies>
     <AndroidTargetFrameworkVersion Condition="'$(AndroidTargetFrameworkVersion)' == ''">v9.0</AndroidTargetFrameworkVersion>
-    <TargetFrameworkVersion>$(AndroidTargetFrameworkVersion)</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v9.0</TargetFrameworkVersion>
```
Comment (samhouts) [inline Xamarin.Forms.Platform.Android/Xamarin.Forms.Platform.Android.csproj]: Need to revert this

File: Xamarin.Forms.Platform.UAP/FormsButton.cs [hunk 8]
```diff
@@ -73,11 +75,13 @@ void UpdateBackgroundColor()
...
 		{
+			var radius = BorderRadius == -1 ? 0 : BorderRadius;
+
 			if (_contentPresenter != null)
-			{
-				var radius = BorderRadius == -1 ? 0 : BorderRadius;
 				_contentPresenter.CornerRadius = new Windows.UI.Xaml.CornerRadius(radius);
-			}
+
+			if (_rootGrid != null)
```
Comment (samhouts) [inline Xamarin.Forms.Platform.UAP/FormsButton.cs]: This should have the added benefit of getting rid of the square highlight behind Buttons with rounded corners. Should look to see if there are any open issues about this on UWP.

File: Xamarin.Forms.Platform.iOS/Renderers/FloatingActionButtonRenderer.cs [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+		{
+			base.OnElementChanged(e);
+
+			if (e.NewElement != null)
+			{
+				if (Control == null)
+				{
+					SetNativeControl(CreateNativeControl());
+
+					Debug.Assert(Control != null, "Control != null");
```
Comment (samhouts) [inline Xamarin.Forms.Platform.iOS/Renderers/FloatingActionButtonRenderer.cs]: Probably don't need this

File: Xamarin.Forms.Platform.UAP/FloatingActionButtonRenderer.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+		{
+			base.OnElementChanged(e);
+
+			if (e.NewElement != null)
+			{
+				if (Control == null)
+				{
+					var button = new FormsButton();
+
+					button.Click += OnButtonClick;
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Platform.UAP/FloatingActionButtonRenderer.cs]: You can unsubscribe the event when doing the dispose.

File: Xamarin.Forms.ControlGallery.iOS/AppDelegate.cs (hunk 5 above)
Comment (samhouts) [reply Xamarin.Forms.ControlGallery.iOS/AppDelegate.cs]: seems fine, i'll leave it

**Comment by samhouts** [comment]
@jsuarezruiz I think we've already taken over this one. :)

File: Xamarin.Forms.Controls/CoreGalleryPages/FloatingActionButtonCoreGalleryPage.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+		}
+
+		protected override bool SupportsFocus
+		{
+			get { return false; }
+		}
+
+		protected override void InitializeElement(FloatingActionButton element)
+		{
+			element.HorizontalOptions = LayoutOptions.Center;
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Controls/CoreGalleryPages/FloatingActionButtonCoreGalleryPage.cs]: None of the specific FloatingActionButton properties are used here. Can we add ImageSource, Size and Command here?
```
Test. FloatingActionButton.ImageSource
Test. FloatingActionButton.Size
Test. FloatingActionButton.Command
```

File: Xamarin.Forms.Controls/GalleryPages/DynamicViewGallery.cs [hunk 12]
```diff
@@ -21,11 +21,12 @@ public class DynamicViewGallery : ContentPage
 						}
 					}, null) },
 				{ nameof(ProgressBar), (() => new ProgressBar(), null) },
+				{ nameof(FloatingActionButton), (() => new FloatingActionButton { ImageSource = "bank.png", HorizontalOptions = LayoutOptions.Center }, null) },
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Controls/GalleryPages/DynamicViewGallery.cs]: Testing I found problems changing dynamically the size of the button.
floatactionbutton-error1
When changing any property of the button (for example in the gif, the FlowDirection property) updates the rendering being correct.

File: Xamarin.Forms.Controls/XamStore/Views/DemoShellPage.xaml [hunk 13]
```diff
@@ -102,32 +109,41 @@
...
+                    <FloatingActionButton  VerticalOptions="End" HorizontalOptions="Center" StyleClass="PopupMenuButton"
+                            BackgroundColor="Orange"
+                            ImageSource="{StaticResource musicIcon}" />
+                    <FloatingActionButton  VerticalOptions="End" HorizontalOptions="Center" StyleClass="PopupMenuButton"
+                            BackgroundColor="Green"
+                            ImageSource="{StaticResource octocatIcon}" />
+                    <FloatingActionButton  VerticalOptions="End" HorizontalOptions="Center" StyleClass="PopupMenuButton, Purple"
+                            ImageSource="{StaticResource gameIcon}" />
+                </StackLayout>
+                <FloatingActionButton VerticalOptions="End" x:Name="TogglePopupButton" HorizontalOptions="Center"
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Controls/XamStore/Views/DemoShellPage.xaml]: Very nice sample! :heart:
Captura de pantalla 2019-08-16 a las 14 35 59

File: Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+using Android.Graphics.Drawables;
+using Android.Widget;
+
+namespace Xamarin.Forms.Platform.Android
+{
+	public class FloatingActionButtonRenderer : AFloatingButton,
+		IVisualElementRenderer, IViewRenderer, ITabStop,
+		AView.IOnAttachStateChangeListener, AView.IOnFocusChangeListener, AView.IOnClickListener, AView.IOnTouchListener
+	{
+		const float SmallSize = 44;
```
Comment (jsuarezruiz) [inline Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs]: We have these constants in each renderer with the same values. I would move them to the abstraction.

Comment (PureWeen) [inline Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs]: Is 44 the right size?  Should it be 40?
https://developer.android.com/reference/com/google/android/material/floatingactionbutton/FloatingActionButton.html#SIZE_MINI

Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs]: I have moved these const values directly to the enumeration that defines the sizes.

File: Xamarin.Forms.Controls/CoreGalleryPages/FloatingActionButtonCoreGalleryPage.cs (hunk 11 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Controls/CoreGalleryPages/FloatingActionButtonCoreGalleryPage.cs]: Added these properties!

File: Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs (hunk 14 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.Android/Renderers/FloatingActionButtonRenderer.cs]: Yes, you're right, Shane: https://material.io/design/components/buttons-floating-action-button.html#anatomy
FAB containers come in two sizes:
- Default (56 x 56dp)
- Mini (40 x 40dp)

I have applied changes in sizes.

File: Xamarin.Forms.Platform.iOS/Renderers/FloatingActionButtonRenderer.cs [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+			if (Element == null)
+				return SizeF.Empty;
+
+			if (Element.Size == FloatingActionButtonSize.Mini)
+				return new SizeF(SmallSize, SmallSize);
+
+			return new SizeF(NormalSize, NormalSize);
+		}
+
+		public bool IsDisposed { get; private set; }
```
[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.iOS/Renderers/FloatingActionButtonRenderer.cs]: No. I have made changes. We use this property when implementing `IImageVisualElementRenderer`. It is not necessary at public level.

**Comment by samhouts** [comment]
@sichy Thank you so much for starting this PR. This was a good lesson for us that we didn't quite get the specification right the first time. Per @PureWeen's comments (https://github.com/xamarin/Xamarin.Forms/pull/3295#pullrequestreview-277503026), we have some further work to do before we can really do a proper implementation. I'm going to go ahead and close this PR now.

**Pull request closed without merging**